apps/rabbitmq_prelaunch/src/rabbit_boot_state.erl

```diff
@@ -0,0 +1,97 @@
+%%%-------------------------------------------------------------------
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at https://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2007-2020 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(rabbit_boot_state).
+
+-include_lib("eunit/include/eunit.hrl").
+
+-export([get_boot_state/0,
+         set_boot_state/1,
+         wait_for_boot_state/1]).
+
+-define(PT_KEY_BOOT_STATE,    {?MODULE, boot_state}).
+
+-type boot_state() :: 'stopped' | 'booting' | 'ready' | 'stopping'.
+
+-export_type([boot_state/0]).
+
+-spec get_boot_state() -> boot_state().
+get_boot_state() ->
+  persistent_term:get(?PT_KEY_BOOT_STATE, stopped).
+
+-spec set_boot_state(boot_state()) -> term().
+set_boot_state(BootState) ->
+  rabbit_log_prelaunch:debug("Change boot state to `~s`", [BootState]),
+  ?assert(is_boot_state_valid(BootState)),
+  case BootState of
+    stopped -> persistent_term:erase(?PT_KEY_BOOT_STATE);
+    _ -> persistent_term:put(?PT_KEY_BOOT_STATE, BootState)
+  end,
+  notify_boot_state_listeners(BootState).
+
+-spec wait_for_boot_state(boot_state()) -> term().
+wait_for_boot_state(BootState) ->
+  wait_for_boot_state(BootState, infinity).
+
+wait_for_boot_state(BootState, Timeout) ->
+  ?assert(is_boot_state_valid(BootState)),
+  wait_for_boot_state1(BootState, Timeout).
+
+wait_for_boot_state1(BootState, infinity = Timeout) ->
+  case is_boot_state_reached(BootState) of
+    true  -> ok;
+    false -> wait_for_boot_state1(BootState, Timeout)
```

Comment:
@dumbbell is this intended to wait in a tight loop? Also, since the only public wait_for uses a wait of infinity, it seems we will always follow this path. Should we not just expose the function with timeout, and invoke with the BOOT_FINISH_TIMEOUT when this is called in rabbit.erl?

Reply:
Oops, you're right, it lacks a `timer:sleep(...)` like the clause with an integer timeout.

I agree for the use of `?BOOT_FINISH_TIMEOUT`.